pippy/ModelSplit.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import logging
+from typing import Dict, List, Tuple
+
+import torch
+import pippy.fx
+from pippy.IR import pipe_split
+
+"""
+Analyze size of parameters/buffers used by each node in the graph
+Here node can be a `call_function` or a `call_module`
+Returns a Dict that uses Node as key, where value is another Dict that maps from parameter name of that Node to the
+size of that parameter
+"""
+def _analyze_node_size(gm: pippy.fx.GraphModule) -> Dict[pippy.fx.Node, Dict[str, int]]:
+    # state_dict helps us to get parameter sizes
+    state_dict = gm.state_dict()
+
+    # Function Parameter Usage
+    node_param_sizes : Dict[pippy.fx.Node, Dict[str, int]] = {}
+    for node in gm.graph.nodes:
+        if node.op == 'get_attr':   # a parameter node
+            param_name = node.target
+            assert param_name in state_dict
+            param = state_dict[param_name]
+            # Find use site of this parameter
+            for user in node.users:
+                param_sizes = node_param_sizes.setdefault(user, {})
+                param_sizes.setdefault(param_name, param.numel())
+
+    # Module Parameter Usage
+    for node in gm.graph.nodes:
+        # We calcuate size of a user-defined submodule as a whole
+        if node.op == 'call_module':
+            param_sizes : Dict[str, int] = {}
+            submod: torch.nn.Module = gm.get_submodule(node.target)
+            for param_name, param in submod.named_parameters():
+                param_sizes.setdefault(param_name, param.numel())
```

Comment:
Should we mulitply by the size of the element (e.g. number of bytes ?)

Reply:
Hi @aazzolini that's a good suggestion. I hesitated between byte-based and element-based too when designing the API. I eventually went with the element-based solution because it is more in line with other torch APIs (e.g. `Tensor.size()`). Also more in line with how we call those models, like T5-3b, which means 3 billion parameters. Please feel free to let me know if you have other concerns.

Reply:
That makes sense, no concerns. This looks good.